### ISSUE TITLE
Implement a first working version with a stable API

### ISSUE THREAD

**Issue by jpommerening** [LaxarJS/laxar-vue-adapter#1]
Get this adapter ready for production.

Current state: Working, untested.

/cc @x1B 🍾

**Comment by x1B** [comment]
Hot stuff!

**Comment by jpommerening** [comment]
TBD:
- [x] How do we pass injections to the Vue component? (& when)
- [x] Can we somehow cache template compilation or make sure we do it just once per artifact?  
  (Also, the current implementation defers controller creation until `domAttachTo` is called because we have to wait for the template HTML in case the the widget uses a separate file…)
- [x] Controls should probably be registered automatically via the `components` [option](http://vuejs.org/api/#components). It might even be possible to achieve interop with controls instantiated by other adapters.

**Comment by jpommerening** [comment]
The [`vue-loader`](https://vue-loader.vuejs.org/) is pretty awesome. It can hot-reload both CSS and HTML while keeping the state of the component. If I change the code to wrap the widget with a [`<component>`](http://vuejs.org/guide/components.html#Dynamic-Components) it [might](https://github.com/vuejs/vue-hot-reload-api/blob/2e1ee4464e65191195e6a5de5ab95ce4108c015c/index.js#L119) even reload the controller.

**Comment by x1B** [comment]
@jpommerening I think the crucial part is getting the controller instantiated before the DOM attached. Otherwise, Vue widgets might be missing important events. Regarding injections, maybe the [react-adapter](https://github.com/LaxarJS/laxar-react-adapter) could be an inspiration?

**Comment by jpommerening** [comment]
Depends on LaxarJS/laxar#390.

### Injections

``` vue
<template>
   <label>
      <!-- since axContext is mapped to the `$data` property, widget features can be used seamlessly -->
      <span v-html="features.label.htmlText"></span>
      <input v-bind:value="message"></input>
   </label>
</template>

<script>
export default {
   data() {
      return {
         message: "Hello!"
      };
   },
   injections: [ 'axGlobalLog', 'axContext' ],
   created() {
      // Injections available as items in `this.$injections` in the order they were requested:
      const [ log, context ] = this.$injections;

      // "Named" injections are available on `this.$options.injections`
      log === this.$options.injections.axGlobalLog;
      // `axContext` properties are also available via Vue.js' this.$data
      context.eventBus === this.$data.eventBus;
   }
};
</script>
```

### Template compilation

Ideally, templates are compiled by the `vue-loader`. If they are compiled during runtime the "standalone" build of Vue.js is required and templates will be compiled before components are attached to the DOM.

### Controls

Controls can be "requested" with the `"controls"` property in `widget.json` and `control.json` files. They are [locally registered](https://vuejs.org/v2/guide/components.html#Local-Registration) as components with the name specified in the `control.json`.

**Comment by jpommerening** [comment]
Hot reload of templates and styles embedded in the main widget source worked out of the box. To enable it for themed templates and styles these need to be processed by the `vue-loader` (see [README](https://github.com/LaxarJS/laxar-vue-adapter#precompiling-themed-assets)). The adapter then attaches the `beforeCreate` and `beforeDestroy` callbacks from the compiled template to the widget component instance.

**Comment by jpommerening** [comment]
Added support for injections in controls & refactored the implementation of injections to use Vue mixins.
Controls can inject the same services as AngularJS controls including `axWidgetServices` (even the implementation is quite similar).

**Comment by x1B** [comment]
Assigning to me for now, to ensure compatibility with laxar-mocks and to get a first version into the registry.

**Comment by jpommerening** [comment]
@x1B: Regarding widget areas… Vue.js already has a similar feature called "[slots](https://vuejs.org/v2/guide/components.html#Content-Distribution-with-Slots)". I think it would be pretty cool if we could piggyback on that feature to implement widget areas. Unfortunately it does not map well onto the hooks Laxar currently exposes, because both Vue and Laxar want to handle the insertion and removal of elements themselves.

**Comment by x1B** [comment]
To be determined:
- maybe rename the class-level field from `injections` to `axInjections`, for future-proofing and in order to avoid confusion
- for now, I suggest sticking _either_ to the named `$options.injections` map, _or_ the `$injections` array on the instance. Probably the former. If using the latter, we might want to rename to `.axInjections`

Update: using a mixin instead, which makes it OK to then use an array called `this.$injections`.

**Comment by x1B** [comment]
Otherwise: very pleasant to work with so far! 🥇

**Comment by x1B** [comment]
I am adding a new injection `vueComponent` which simply references the widget component instance itself.
This is intended for use in tests because otherwise tests would not be able to access the component, making "grey box" tests (that check the widget model rather than the DOM)

An alternative would be trying to use the `__vue__` property to access the component from the DOM, but that is not official API, and [apparently contended](https://github.com/vuejs/vue/issues/4957). Also, this would require to always render the component for grey-box testing.

**Comment by x1B** [comment]
@jpommerening I'll grab this one (again), as you have enough on your plate with LaxarJS/laxar#403 :)

**Comment by jpommerening** [comment]
@x1B Go ahead!

If you're concerned about the `injections` option trying to look like something Vue brings us, we could easily expose the injections mixin, or make it a Vue plugin so using it is a bit more explicit:

```js
import Vue from 'vue';
import { Injections } from 'laxar-vue-adapter';

// either
Vue.use(Injections);

export default {
   injections: [ 'axEventBus' ]
};

// or
export default {
   mixins: [ Injections ],
   injections: [ 'axEventBus' ]
};
```

Now, back to LaxarJS/laxar#403 ;)

**Comment by x1B** [comment]
Now, mixins is an interesting concept (`Vue.use` affects the entire app, which I don't like). But I also really like that a vue-widget can simply export a component without doing any importing at all. 

Still, I'll have closer look at the mixin idea, since custom injections are not needed all that often.

**Comment by jpommerening** [comment]
If we go the mixin route, we could also parametrise the mixin with the requested injections! (`mixins: [ Injections( 'axEventBus', … ) ]`)

Also, I found my code for vue widget areas. I'll open an issue to discuss that at later point.

**Comment by x1B** [comment]
Nice! wrt widget areas: good to know, was already thinking about those. But that's another issue.

**Comment by x1B** [comment]
Implemented on `master` (`v2.0.0`).

### Upgrade Information

Users of alpha.2 and earlier must change their widget components when upgrading, at least if those components are using injections.
Those now have to use the `injections` mixin (see README), and the injected services must be accessed through the `$injections` array.

**Comment by x1B** [comment]
Kudos to @jpommerening for the initial work on this 🏆
